### PR TITLE
MODEUR-14: book report

### PR DESCRIPTION
**Pull request by julianladisch**

### CODE REVIEW

File: src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java [hunk 1]
```diff
@@ -1109,21 +1109,27 @@ static String getRanges(JsonArray ar, String startProp, String endProp) {
 
   Future<Void> getUseOverTime(Vertx vertx, RoutingContext ctx) {
     String format = ctx.request().params().get("format");
+    boolean journal;
```
Comment (MikeTaylor) [inline src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java:1112]: For what it's worth, when I see a variable called `journal`, my immediate assumption is that it holds a reference to a journal. For a boolean that says whether we're interested in journals or not, I would be more likely to use `isJournal` or similar.

File: src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java [hunk 2]
```diff
@@ -1169,26 +1175,36 @@ static String getRanges(JsonArray ar, String startProp, String endProp) {
       LongAdder [] uniqueItemRequestsByPeriod = LongAdder.arrayOfLength(accessCountPeriods.size());
       rowSet.forEach(row -> {
         JsonArray accessCountsByPeriod = new JsonArray();
+        LongAdder accessCountTotal = new LongAdder();
         boolean unique = "Unique_Item_Requests".equals(row.getString("metrictype"));
+        int pos0 = journal ? 6 : 5;
```
Comment (MikeTaylor) [inline src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java:1180]: I missed the `7` in the previous version, but these magic numbers make me nervous. Can we make them named constants?

File: src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java [hunk 3]
```diff
@@ -1224,28 +1239,26 @@ static String getRanges(JsonArray ar, String startProp, String endProp) {
    *   WHERE daterange($3, $4) @> lower(usageDateRange) AND openAccess
    *   GROUP BY titleEntryId
    * ) t1 ON t1.titleEntryId = title_entries.id
-   * WHERE agreementId = $1;
+   * WHERE agreementId = $1
+   *   AND NOT (printISSN IS NULL AND onlineISSN IS NULL)
```
Comment (MikeTaylor) [inline src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java:1243]: Surely this comment is wrong in the BOOK case?

File: src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java [hunk 4]
```diff
@@ -1224,28 +1239,26 @@ static String getRanges(JsonArray ar, String startProp, String endProp) {
...
+        .append(journal ? " printISSN, onlineISSN, "
+                        : " ISBN, ")
         .append(openAccess ? "'OA_Gold' AS accessType, "
                            : "'Controlled' AS accessType, ")
         .append(unique ? "'Unique_Item_Requests' AS metricType, "
                        : "'Total_Item_Requests' AS metricType, ");
     for (int i = 0; i < periods; i++) {
       if (i > 0) {
-        sql.append(" + ");
+        sql.append(", ");
```
Comment (MikeTaylor) [inline src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java:1258]: Is this fixing a straight-up mistake?

File: src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java [hunk 5]
```diff
@@ -1261,16 +1274,9 @@ private static void useOverTimeJournal(StringBuilder sql, TenantPgPool pool,
...
-  Future<Void> getUseOverTimeBook(Vertx vertx, RoutingContext ctx) {
-    // Work in Progress (WIP), hardcoded example JSON
-    ctx.response().setStatusCode(200);
-    ctx.response().putHeader("Content-Type", "application/json");
-    ctx.response().end(ResourceUtil.load("/openapi/examples/report.json"));  // FIXME
-
-    return Future.succeededFuture();
+    sql.append(" WHERE agreementId = $1 AND ")
+        .append(journal ? "NOT " : "")
+        .append("(printISSN IS NULL AND onlineISSN IS NULL)");
```
Comment (MikeTaylor) [inline src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java:1279]: Some question as above.

**Review by MikeTaylor — APPROVED** [review]
Looks good!

File: src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java (hunk 1 above)
Comment (julianladisch) [reply src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java]: Changed.

File: src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java (hunk 2 above)
Comment (julianladisch) [reply src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java]: Changed.

File: src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java (hunk 3 above)
Comment (julianladisch) [reply src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java:1245]: This comment lists _one_ example, this is sufficient.
The SQL changes when changing the number of periods (n0, n1, n2, ...), openAccess (true/false), or report (journal, book).

File: src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java (hunk 4 above)
Comment (julianladisch) [reply src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java:1260]: No, it is a misleading diff created by GitHub.
This column has been removed, see example in (old) line 1211:
```n0 + n1 AS total,```

File: src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java (hunk 5 above)
Comment (julianladisch) [reply src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java:1281]: Generated SQL

- for book: `AND NOT (printISSN IS NULL AND onlineISSN IS NULL)`
- for journal: `AND (printISSN IS NULL AND onlineISSN IS NULL)`

This is correct and tested by unit tests.

File: src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java (hunk 1 above)
Comment (MikeTaylor) [reply src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java]: Thanks!

File: src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java (hunk 2 above)
Comment (MikeTaylor) [reply src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java]: Appreciated.

File: src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java (hunk 3 above)
Comment (MikeTaylor) [reply src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java:1245]: Gotcha.

File: src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java (hunk 4 above)
Comment (MikeTaylor) [reply src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java:1260]: OK.

File: src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java (hunk 5 above)
Comment (MikeTaylor) [reply src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java:1281]: Ah, yes, I see. My bad.

**Comment by MikeTaylor** [comment]
Thanks for this, great to have it merged.

**Pull request merged**